src/controls/ButtonControls.cpp

```diff
@@ -0,0 +1,185 @@
+#include <esp_sleep.h>
+#include <driver/rtc_io.h>
+#include <driver/gpio.h>
+#include <esp32/ulp.h>
+#include <esp_log.h>
+#include <esp_timer.h>
+#include "ButtonControls.h"
+#include "ulp_main.h"
+
+// 100 ms debounce on the buttons
+const int BUTTON_DEBOUNCE = 1000;
+
+IRAM_ATTR void button_interrupt_handler(void *param)
+{
+  Button *button = (Button *)param;
+  button->handle_interrupt();
+}
+
+Button::Button(gpio_num_t gpio_pin, int active_level, ButtonCallback_t callback)
+    : gpio_pin(gpio_pin),
+      active_level(active_level),
+      callback(callback)
+{
+  // setup the pin
+  gpio_set_direction(gpio_pin, GPIO_MODE_INPUT);
+  gpio_set_pull_mode(gpio_pin, active_level == 0 ? GPIO_PULLUP_ONLY : GPIO_PULLDOWN_ONLY);
+  // wire up the interrupt handler
+  gpio_isr_handler_add(gpio_pin, button_interrupt_handler, this);
+  // start off waiting for the button to be pressed - edge interrupts seem to be flaky so we
+  // have to be a bit more clever
+  button_pressed = false;
+  gpio_set_intr_type(gpio_pin, active_level == 0 ? GPIO_INTR_LOW_LEVEL : GPIO_INTR_HIGH_LEVEL);
+  gpio_intr_enable(gpio_pin);
+}
+
+void Button::handle_interrupt()
+{
+  if (button_pressed)
+  {
+    // the button was pressed - so now it must have been released
+    button_pressed = false;
+    int64_t button_release_time = esp_timer_get_time();
+    // has enough time passed to call this a button press?
+    if (button_release_time - button_press_start > BUTTON_DEBOUNCE)
+    {
+      // call the callback
+      callback();
+    }
+    // start listening for the button to be pressed again
+    gpio_set_intr_type(gpio_pin, active_level == 0 ? GPIO_INTR_LOW_LEVEL : GPIO_INTR_HIGH_LEVEL);
+  }
+  else
+  {
+    // the button was not pressed - so now it must be pressed
+    button_pressed = true;
+    button_press_start = esp_timer_get_time();
+    // start listening for the button to be released
+    gpio_set_intr_type(gpio_pin, active_level == 0 ? GPIO_INTR_HIGH_LEVEL : GPIO_INTR_LOW_LEVEL);
+  }
+}
+
+extern const uint8_t ulp_main_bin_start[] asm("_binary_ulp_main_bin_start");
+extern const uint8_t ulp_main_bin_end[] asm("_binary_ulp_main_bin_end");
+
+ButtonControls::ButtonControls(
+    gpio_num_t gpio_up,
+    gpio_num_t gpio_down,
+    gpio_num_t gpio_select,
+    int active_level,
+    ActionCallback_t on_action)
+    : gpio_up(gpio_up), gpio_down(gpio_down), gpio_select(gpio_select),
+      active_level(active_level),
+      on_action(on_action)
+{
+  gpio_install_isr_service(0);
```

Comment:
Note that Touch already does this before, so you should get a warning here!

Reply:
We have to research on this one. If it’s just a warning and does not reset wouldn’t be so bad anyways